tox.ini

```
@@ -63,4 +63,14 @@ deps =
 commands =
   black {posargs:src tests tools}
 passenv =
-  {[testenv:common]passenv}
\ No newline at end of file
+  {[testenv:common]passenv}
+[testenv:{,py310-,py311-,py312-}pytest]
+description = Run pytest
+setenv =
+    PYTHONPATH = {toxinidir}/src
+deps =
+  pytest
+commands =
+  pytest
```

```suggestion
  pytest {posargs:tests --junitxml=./tests_report.xml}
```
This syntax means, that if you just run:
`tox -e pytest`
then it will call by default what is to the right from `posargs:`, so,
`pytest tests --junitxml=./tests_report.xml`
By this you make sure that tests are discovered only in tests directory and you will also get a nice XML report with results.
And it gives you flexibility to pass extra arguments if you want to modify pytest invocation.
For example you may want to run only one test with capturing print statements (`-s`) and you can do it then by:
`tox -e pytest -- tests/example_test.py::test_is_prime -s`
This `--` double dash means that tox will pass over arguments to its interior `posargs`.
Very handy in long run.